crates/stackable-versioned-macros/src/gen/venum/mod.rs

```diff
@@ -0,0 +1,182 @@
+        let mut versions = self.versions.iter().peekable();
+
+        while let Some(version) = versions.next() {
+            token_stream.extend(self.generate_version(version, versions.peek().copied()));
+        }
+
+        token_stream
+    }
+}
+
+impl VersionedEnum {
+    fn generate_version(
+        &self,
+        version: &ContainerVersion,
+        next_version: Option<&ContainerVersion>,
+    ) -> TokenStream {
+        let mut token_stream = TokenStream::new();
+        let enum_name = &self.ident;
+
+        // Generate variants of the enum for `version`.
+        let variants = self.generate_enum_variants(version);
+
+        // TODO (@Techassi): Make the generation of the module optional to
+        // enable the attribute macro to be applied to a module which
+        // generates versioned versions of all contained containers.
+
+        let deprecated_attr = version.deprecated.then_some(quote! {#[deprecated]});
+        let module_name = &version.ident;
+
+        // Generate tokens for the module and the contained struct
+        token_stream.extend(quote! {
+            #[automatically_derived]
+            #deprecated_attr
+            pub mod #module_name {
+                pub enum #enum_name {
+                    #variants
+                }
+            }
+        });
+
+        // Generate the From impl between this `version` and the next one.
+        if !self.skip_from && !version.skip_from {
+            token_stream.extend(self.generate_from_impl(version, next_version));
+        }
+
+        token_stream
+    }
+
+    fn generate_enum_variants(&self, version: &ContainerVersion) -> TokenStream {
+        let mut token_stream = TokenStream::new();
+
+        for variant in &self.items {
+            token_stream.extend(variant.generate_for_container(version));
+        }
+
+        token_stream
+    }
+
+    fn generate_from_impl(
+        &self,
+        version: &ContainerVersion,
+        next_version: Option<&ContainerVersion>,
+    ) -> TokenStream {
+        if let Some(next_version) = next_version {
+            let next_module_name = &next_version.ident;
+            let module_name = &version.ident;
+
+            let from_ident = &self.from_ident;
+            let enum_ident = &self.ident;
+
+            let mut variants = TokenStream::new();
+
+            for item in &self.items {
+                variants.extend(item.generate_for_from_impl(
+                    module_name,
+                    next_module_name,
+                    version,
+                    next_version,
+                    enum_ident,
+                ))
+            }
+
+            // TODO (@Techassi): Be a little bit more clever about when to include
+            // the #[allow(deprecated)] attribute.
+            return quote! {
+                #[automatically_derived]
+                #[allow(deprecated)]
+                impl From<#module_name::#enum_ident> for #next_module_name::#enum_ident {
```

Comment:
Can we conditionally include it if there is a deprecation, like what you did further above?

```suggestion
            let allow_deprecated_attr = version.deprecated.then_some(quote! {#[allow(deprecated)]});
            return quote! {
                #[automatically_derived]
                #allow_deprecated_attr
                impl From<#module_name::#enum_ident> for #next_module_name::#enum_ident {
```

Reply:
Yes, that is the idea. But sadly it is not as trivial as in the above case. We need to take into account when the `From` implementation converts *to* as version which includes a deprecated field or variant.

Reply:
I would like to add that feature in a follow-up PR. Any objections @NickLarsenNZ?

Reply:
Yeah sure. Was just an idea without thinking about the complexity of it.